pkg/cmd/factory/remote_resolver.go

```diff
@@ -0,0 +1,80 @@
+package factory
+
+import (
+	"errors"
+	"net/url"
+	"sort"
+
+	"github.com/cli/cli/context"
+	"github.com/cli/cli/git"
+	"github.com/cli/cli/internal/config"
+	"github.com/cli/cli/internal/ghinstance"
+)
+
+type remoteResolver struct {
+	readRemotes   func() (git.RemoteSet, error)
+	getConfig     func() (config.Config, error)
+	urlTranslator func(*url.URL) *url.URL
+}
+
+func (rr *remoteResolver) Resolver() func() (context.Remotes, error) {
+	var cachedRemotes context.Remotes
+	var remotesError error
+
+	return func() (context.Remotes, error) {
```

Comment:
The caching here feels premature. Unless I'm missing something the costliest action here is reading remotes on disk and unless we've benchmarked that as human-noticeable I'd rather get rid of the complexity of this closure-based approach and not worry about caching remotes.

Reply:
It might be, and you're right that I just guessed that this might be slow.

This logic includes shelling out to a subprocess (`git remote -v`) and parsing that and `~/.ssh/config` if it exists. Now that I've actually measured it on my machine (high-end MBP), it usually takes 6-10ms. However, it does get called at least twice for some commands, and I have a hunch some code paths around `pr create` and `pr checkout` that consult remotes a lot might call it even more than twice. That puts the combined cost of the lookup in the 20–40ms range, which I think starts being noticeable by humans. However, in most cases this caching only shaves off up to 10ms, so you might be right that the caching is premature.

What are your thoughts now that we know the numbers?

Reply:
I was envisioning that this would only be called 1-2 times per command; but good point that it could be a number of times for pr create. It sounds like this can creep into the human-noticeable range so let's keep it for now; it's easy enough to take out later.